### ISSUE TITLE
Differentiate between indexed and non-indexed dynamic-array event fields 

### ISSUE THREAD

**Issue by davekaj** [graphprotocol/graph-node#913]
Right now graph-node can properly distinguish between indexed and non indexed dynamic arrays (arrays, strings, bytes) when indexing an event. (I believe this is rooted in parity eth-abi)

The outlier problem is this - when we start to track generic interface events, i.e. not supplying a contract address, there is ambiguity we have to deal with. (see https://github.com/ensdomains/resolvers/issues/13#issuecomment-488829262 )

Example:
` event TextChanged(bytes32 indexed node, string indexed indexedKey, string key); `
` event TextChanged(bytes32 indexed node, string indexedKey, string key); `

The above two events have the exact same event signature, however the data they return is different. The top one returns `string indexed indexedKey` as `bytes32` (because the string is hashed and stored as such), and the second one `string indexedKey` is handled as a string - which makes multiple `bytes32` to represent the variable length. 

The way graph-node is implemented now, it can only deal with data type for the event field. We need to allow it to support both. This is a minor blocker on ENS, and an edge case, but it is worthy of us to support it. 

Suggestions:
- look at the ABI for the event and determine if it is `indexed: true` or `false`. Then , combine that with the event signature, so that we cab distinguish them from each other

**Comment by Arachnid** [comment]
Some way to pick one and gracefully handle errors for the other would be a good workaround here. In the example above, we only care about the former, but don't want errors from the latter to torpedo event decoding.

**Comment by Arachnid** [comment]
(For that matter: What happens if someone just submits an event with the wrong data format? Will it torpedo our entire subgraph? That's easy to do once you're capturing events without a sending address specified.)

**Comment by Jannis** [comment]
The fact that the `indexed` hint doesn't make it into the event topic in Ethereum is unfortunate. It means that
```
event Foo(indexed uint256, indexed uint256)
event Foo(uint256, uint256)
```
both get the same signature, despite their data being encoded differently. Without the right event ABI, you therefore can't decode any event safely.

I think we can solve this though. Let me think through the different areas that affected:

## Event signatures in the manifest

First, we could allow event signatures in the manifest to include `indexed`, e.g. with
```
- eventHandlers:
    - event: Foo(indexed uint256,indexed uint256)
```
That would allow us to distinguish overloaded events from each other. 

We could auto-migrate existing subgraphs to the new signatures as part of `graph codegen`, `graph build` and `graph deploy`. The list of events available in ABIs would need to be updated to prefix indexed params with `indexed`. When validating a subgraph, we'd then fail if an event in the manifest is missing `indexed` hints (due to the auto-migration, this won't happen).

## Decoding incoming events

Decoding events is where the ambiguity kicks in. Let's say we receive an event whose topic matches the above `Foo(indexed uint256,indexed uint256)`. We can't be sure the event is actually _this_ variant or another without `indexed` parameters.

In other words, https://github.com/graphprotocol/ethabi/blob/graph-patches/ethabi/src/event.rs#L122 could fail parsing the event using the `Foo(indexed uint256,indexed uint256)` ABI.

We have different options on how to handle this situation:

1. Fail the subgraph. According to https://github.com/graphprotocol/graph-node/blob/85b9ec73a65f3b95324db6bcc1c2caa15384fb1c/runtime/wasm/src/host.rs#L710-L723, this is our current behavior.
2. Log a warning and skip the event. Clearly, it's not matching our ABI.

If we switch to 2, mismatches in the event data would be handled gracefully.

## Code generation

We already enumerate the classes generated for overloaded event variants, so we're safe here. This part doesn't need changing.

**Comment by Arachnid** [comment]
I just encountered a real-world issue with this on another event. Having the whole subgraph fail when an event can't be decoded means that, for filters that don't specify a source address, effectively anyone can cause your subgraph to fail by publishing an invalid (or just different) event to the chain.

The error message isn't great, either:
```
Subgraph instance failed to run: Failed to process trigger: Failed to parse parameters of event: OwnerChanged(address): Invalid data, code: SubgraphSyncingFailure, id: QmZPYMTQQmeV9jUEg977sB8bjSv4h17NYQA5GrWTU5k2XV, updates: 0
```

There are over 600 transactions in the block that failed; it'd be useful to know which one caused the failure.

As a first step, could you change the node to only issue a warning and skip the event if decoding fails?
